### ISSUE TITLE
Grain matching(glob/pcre) should be able to match wildcards in the middle

### ISSUE THREAD

**Issue by dnd** [saltstack/salt#19651]
Grain glob and grain-pcre matching currently only seem to support wildcards at the tail of the expression.

Given a grain structure of the following:

```yaml
apps:
  blog:
    - web
    - worker
    - db
```

I can match the grain on `apps:blog:*` or I can match `apps:blog:(web|worker)`, but I cannot get say all app workers using either `app:*:worker` or `app:.*:worker`.

The workaround is to flip the index into another set of grains, but it seems like this functionality should be built into salt. 

The grains already understand `:` as a divider between levels, so a wildcard wouldn't imply anything anywhere, it would just need to compare the keys at the given level.

**Comment by rallytime** [comment]
Thanks for the report! We should be able to match that. For completeness, would you please post the output of `salt --versions-report`?

**Comment by dnd** [comment]
```
# salt --versions-report
           Salt: 2014.7.0-1079-gdbbab28
         Python: 2.6.5 (r265:79063, Feb 27 2014, 19:44:14)
         Jinja2: 2.7.3
       M2Crypto: 0.20.1
 msgpack-python: 0.1.9.final
   msgpack-pure: Not Installed
       pycrypto: 2.0.1
        libnacl: Not Installed
         PyYAML: 3.09
          ioflo: Not Installed
          PyZMQ: 13.0.0
           RAET: Not Installed
            ZMQ: 3.2.2
           Mako: 1.0.0
```

**Comment by rallytime** [comment]
Thanks!

**Comment by jagguli** [comment]
Would this need change in both the master and minion or just the master ? Would it be possible to 'extend' the matching functionality, maybe custom matchers?

**Comment by cachedout** [comment]
This has already been implemented on 2016.3 which is already in the release-candidate phase:

```
mp@silver ...devel/salt/tests % sudo salt -G 'apps:*:worker' test.ping                                                                                             
silver:
    True
```
